### PR TITLE
Add message to project

### LINKED ISSUE #20: Add messages to projects
**/classes/:classId/projects/:projectId/messages**
- **GET**: Return all messages of the project
- **POST**: Add a message and deliver a notification (type: `all`, so it gets delivered by mail and as notification) using the notificationService. like it is used here https://github.com/kredde/javascript_seminar_backend/blob/master/src/services/project.service.js#L18

### PR DESCRIPTION
**Pull request by hk-zh**
addMessage and getAllMessages

### CODE REVIEW

File: src/services/project.service.js [hunk 1]
```diff
@@ -46,13 +49,40 @@ const updateProjectById = async (id, teacher, body) => {
...
+  const teacherIds = project && project.classes.map((c) => String(c.teacher));
+  if (teacherIds.indexOf(teacher) < 0 || teacherIds.indexOf(messageBody.to) < 0) {
+    throw new ApiError(httpStatus.FORBIDDEN, 'message invalid');
+  }
+  const message = messageService.createMessage(messageBody);
+  await Project.updateOne({ _id: id }, { $push: { messages: message._id } });
+  const senderNotification = sendMessage(await userService.getUserById(messageBody.to), messageBody.message);
+  const receiverNotification = receiveMessage(await userService.getUserById(messageBody.from), messageBody.message);
+  await notificationService.sendNotification(messageBody.from, senderNotification);
+  await notificationService.sendNotification(messageBody.to, receiverNotification);
```
Comment (kredde) [inline src/services/project.service.js]: i dont think the sender needs to receive a notification?

File: src/utils/notifications.js [hunk 2]
```diff
@@ -31,8 +31,30 @@ const createProject = ({ teacher, teacherClass, otherClass }) => ({
   type: 'all'
 });
 
+const sendMessage = (user, messageBody) => ({
+  title: `You have sent message to ${user.firstName} ${user.lastName}`,
+  text: `${messageBody}`,
+  cta: {
+    text: 'Back',
+    url: `${config.frontendHost}`
```
Comment (kredde) [inline src/utils/notifications.js]: "See the message" would be appropriate here

File: src/services/project.service.js (hunk 1 above)
Comment (hk-zh) [reply src/services/project.service.js]: ok, then remove it

File: src/validations/project.validation.js [hunk 3]
```diff
@@ -28,8 +28,30 @@ const getProject = {
...
+const addMessage = {
+  params: Joi.object().keys({
+    projectId: Joi.string().custom(objectId),
+    classId: Joi.string().custom(objectId)
+  }),
+  body: Joi.object().keys({
+    message: Joi.string().required(),
+    from: Joi.string().custom(objectId),
+    to: Joi.string().custom(objectId),
+    classId: Joi.string().custom(objectId)
```
Comment (kredde) [inline src/validations/project.validation.js:47]: why is this needed in the body?

**Review by kredde — APPROVED** [review]
This works. 
But it would be easier for the frontend to not have to send `to` and `from` as one could extract it using the user from the request.
We can leave it like this for now and change it, if the frontend team wants us to

File: src/validations/project.validation.js (hunk 3 above)
Comment (kredde) [reply src/validations/project.validation.js:47]: @Hongkuan-Zhou ?

Comment (hk-zh) [reply src/validations/project.validation.js:47]: you mean classId?

Comment (hk-zh) [reply src/validations/project.validation.js:47]: we need 'message' and 'to', and as you said 'from' is unnecessary. we leave it like this until frontend team have other requirements. 'classId' is a mistake, i delete it

Comment (hk-zh) [reply src/validations/project.validation.js:47]: i have no permission to directly work on master branch, could you delete it for me ?
